### PR TITLE
Show notifications for new grades

### PR DESCRIPTION
**Pull request by thellmund**
This pull request implements notifications for new grades, as specified in #688. Whenever `DownloadService` is run, `GradesBackgroundUpdater` fetches the user’s grades and compares them with existing grades in `GradesStore`. If any new grades have been added, we display a notification via `GradesNotificationProvider`.

### CODE REVIEW

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/GradesStore.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        val isFirstRefresh: Boolean,
+        val existingGrades: List<String>
+)
+
+class GradesStore @Inject constructor(
+        private val sharedPrefs: SharedPreferences
+) {
+
+    private val isFirstRefresh: Boolean
+        get() = sharedPrefs.contains(KEY_GRADED_COURSES).not()
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/GradesStore.kt]: Yea... no. This is not the best idea. Besides polluting the shared prefs with stuff that should not be in there, it is a rather cumbersome solution for the initialization problem. How about just skipping the notification, if there are more than 5 new grades? Seems a lot simpler and in practice, nobody will receive 5 grades at exactly the same time or within a refresh period. 

Thoughts?

Comment (kordianbruck) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/GradesStore.kt]: (Yes, magic constants and so on, that @pfent will complain about, but its a simpler solution for that problem)

File: app/src/main/java/de/tum/in/tumcampusapp/service/DownloadService.kt [hunk 2]
```diff
@@ -111,7 +115,9 @@ class DownloadService : JobIntentService() {
         val newsSuccess = downloadNews(force)
         val eventsSuccess = downloadEvents()
         val topNewsSuccess = downloadTopNews()
-        return updateNoteSuccess && cafeSuccess && kinoSuccess && newsSuccess && topNewsSuccess && eventsSuccess
+        val gradesSuccess = downloadGrades()
+        return updateNoteSuccess && cafeSuccess && kinoSuccess
+                && newsSuccess && topNewsSuccess && eventsSuccess && gradesSuccess
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/service/DownloadService.kt]: uggh, this class is getting uglier by the number of times we are editing it. We should really write up a todo, to finally fix it.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/GradesStore.kt (hunk 1 above)
Comment (thellmund) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/GradesStore.kt]: Yes, I was not too happy with the initialization problem either. Will update as per your suggestion.

File: app/src/main/java/de/tum/in/tumcampusapp/service/DownloadService.kt (hunk 2 above)
Comment (pfent) [reply app/src/main/java/de/tum/in/tumcampusapp/service/DownloadService.kt]: I tried to break it up a little bit in  #1188.
Really doesn't make a lot of sense having it all in one big service :man_shrugging:

**Pull request merged**